### PR TITLE
[Carry  2337]  shutdown logger after exit

### PR DESCRIPTION
**Pull request by fahedouch**
carry https://github.com/containerd/nerdctl/pull/2337 ( _originally created by @mharwani_)

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
> rebase changes

This shouldn't be a separate commit

**Comment by fahedouch** [comment]
> > rebase changes
> 
> This shouldn't be a separate commit

is it better now ?

File: pkg/logging/logging.go [hunk 1]
```diff
@@ -177,10 +177,11 @@ func getContainerWait(ctx context.Context, hostAddress string, config *logging.C
 	// If task was not found, it's possible that the container runtime is still being created.
 	// Retry every 100ms.
 	ticker := time.NewTicker(100 * time.Millisecond)
+	defer ticker.Stop()
```
Comment (AkihiroSuda) [inline pkg/logging/logging.go:180]: Probably this should be a separate commit

File: pkg/logging/logging.go [hunk 2]
```diff
@@ -207,7 +208,7 @@ func loggingProcessAdapter(ctx context.Context, driver Driver, dataStore, hostAd
 		buf := make([]byte, 32<<10)
 		_, err := io.CopyBuffer(writer, reader, buf)
 		if err != nil {
-			logrus.Errorf("failed to copy stream: %s", err)
+			log.L.Errorf("failed to copy stream: %v", err)
```
Comment (AkihiroSuda) [inline pkg/logging/logging.go:211]: This should meld into the carried commit so that the carried commit is still compilable

File: pkg/logging/logging.go (hunk 1 above)
Comment (fahedouch) [reply pkg/logging/logging.go:180]: fixed

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]
Still contains uncompilable commits

```
$ git checkout a234c8eab60db7b5a0d2933729c9035ae66fa0e4

$ make
GO111MODULE=on CGO_ENABLED=0 GOOS=linux go build -ldflags "-s -w -X github.com/containerd/nerdctl/pkg/version.Version=v1.7.0-10-ga234c8ea -X github.com/containerd/nerdctl/pkg/version.Revision=a234c8eab60db7b5a0d2933729c9035ae66fa0e4"   -o /home/suda/gopath/src/github.com/containerd/nerdctl/_output/nerdctl github.com/containerd/nerdctl/cmd/nerdctl
# github.com/containerd/nerdctl/pkg/logging
pkg/logging/logging.go:179:4: undefined: logrus
make: *** [Makefile:57: nerdctl] Error 1
```

File: pkg/logging/logging.go [hunk 3]
```diff
@@ -189,11 +272,24 @@ func loggerFunc(dataStore string) (logging.LoggerFunc, error) {
...
+			f, err := os.Create(loggerLock)
+			if err != nil {
 				return err
 			}
+			defer f.Close()
+
+			// the logger will obtain an exclusive lock on a file until the container is
+			// stopped and the driver has finished processing all output,
+			// so that waiting log viewers can be signalled when the process is complete.
+			return lockutil.WithDirLock(loggerLock, func() error {
```
Comment (AkihiroSuda) [inline pkg/logging/logging.go:286]: "DirLock" is now misnomer

Comment (fahedouch) [reply pkg/logging/logging.go:286]: Why ? could you be more explicit please?

Comment (AkihiroSuda) [reply pkg/logging/logging.go:286]: loggerLock doesn't seem to be a dir
